File: Scala/problem14.scala

def isEven(n: Int): Boolean = {
  (n % 2) == 0
}


def collatz(n: Int): Stream[Int] = {
  
  def collatzNext(n: Int) = {
    if (isEven(n)) n/2
    else 3*n + 1
  }
  
  if (n > 1) n #:: collatz(collatzNext(n))
  else Stream[Int]()
}

def problem14(limit: Int): Int = {
  
  def newValues(cs: Stream[Int], sols: Map[Int, Int]): List[Int] = {
    
    def isNewVal(n: Int): Boolean = !sols.contains(n)
    // DOESN'T INCLUDE THE NEXT VALUE!
    cs.takeWhile(isNewVal).toList
  }
  
  def newAnswers(cs: List[Int], sols: Map[Int, Int]): Map[Int, Int] = {
    
    lazy val addLen: Int = sols(cs.last)
    def withCounts: List[Int] = cs.reverse.zipWithIndex.tail
    
    withCounts.map(t => (t._1, t._2 + addLen)).toMap
  }
  
  lazy val solutions = scala.collection.mutable.Map[Int, Int](1 -> 1)
  for (n <- (2 until limit)) {
    solutions ++= newAnswers(newValues(collatz(n), solutions), solutions)
  }
  
  solution.maxBy(_._2)._1
}

File: python/problem14.py
def is_even(n):
   return ((n % 2) == 0)


def collatz(n):
    yield n
    while n > 1:
        if is_even(n):
            n = n/2
        else:
            n = 3*n + 1
        yield n


def collatz_length(cs, sols):
    i = 0
    for n in cs:
        if n in sols:
            return i + sols[n]
        i += 1

    return n

    
def problem14(limit):
    
    sols = {1: 1}
    
    for n in range(2, limit):
        collatz_seq = collatz(n)
        sols[n] = collatz_length(collatz_seq, sols)
    
    return max(sols.iterkeys(), key=(lambda key: sols[key]))
